typing/typecore.ml

```diff
@@ -4833,14 +4833,38 @@ let report_error env ppf = function
       fprintf ppf "Variable %s must occur on both sides of this | pattern"
         (Ident.name id);
       spellcheck_idents ppf id valid_idents
-  | Expr_type_clash (trace, explanation) ->
+  | Expr_type_clash (trace, explanation, exp) ->
       report_unification_error ppf env trace
         ~type_expected_explanation:
           (report_type_expected_explanation_opt explanation)
         (function ppf ->
            fprintf ppf "This expression has type")
         (function ppf ->
-           fprintf ppf "but an expression was expected of type")
+           fprintf ppf "but an expression was expected of type");
+
+      let hints elt =
+        match elt, exp with
+        | Ctype.Unification_trace.(Diff
+            { expected = { t = { desc = Tconstr (p2, [], _); _ }; _ } }),
+          Some (Texp_constant (Const_int n)) ->
+            let hint_suffix c =
+              fprintf ppf "@,@[Hint: Did you mean `%d%c'?@]" n c
+            in
+            if Path.same p2 Predef.path_int32 then
+              hint_suffix 'l'
+            else if Path.same p2 Predef.path_int64 then
+              hint_suffix 'L'
+            else if Path.same p2 Predef.path_nativeint then
+              hint_suffix 'n'
+            else if Path.same p2 Predef.path_float then
+              hint_suffix '.'
+            else
```

Comment:
This `else` is not necessary.

Reply:
I would move that case-distinction code into an auxiliary function, for clarity of the code. `literal_prefix_for_type <path>`.

Reply:
I think the whole `hint` function could be moved to a `report_literal_type_constraint` function in order to share it between patterns and expressions.